### PR TITLE
L2 ageing

### PR DESCRIPTION
**Pull request by toanju**
bridging entries are now no longer timing out despite having traffic on them

### CODE REVIEW

File: src/netlink/cnetlink.cpp [hunk 1]
```diff
@@ -282,6 +300,114 @@ void cnetlink::nl_cb_v2(struct nl_cache *cache, struct nl_object *old_obj,
...
+      VLOG(2) << __FUNCTION__ << ": ifindex=" << ifindex
+              << ", bridge=" << bridge << ", br_link=" << OBJ_CAST(br_link);
+
+      if (br_link) {
+        // learn the source mac
+        bridge->learn_source_mac(br_link, p.pkt);
+      }
+    }
+
+    VLOG(2) << __FUNCTION__ << ": send pkt " << p.pkt
```
Comment (DFritzsche) [inline src/netlink/cnetlink.cpp]: not going to stress this, but technically speaking the packets are not sent to the kernel, just queued into the tap_man

File: src/of-dpa/controller.cpp [hunk 2]
```diff
@@ -561,14 +462,23 @@ int controller::l2_addr_remove_all_in_vlan(uint32_t port,
...
 int controller::l2_addr_add(uint32_t port, uint16_t vid,
-                            const rofl::cmacaddr &mac, bool filtered) noexcept {
+                            const rofl::caddress_ll &mac, bool filtered,
+                            bool permanent) noexcept {
   int rv = 0;
   try {
     rofl::crofdpt &dpt = set_dpt(dptid, true);
+
+    if (!permanent)
+      fm_driver.set_idle_timeout(300);
```
Comment (DFritzsche) [inline src/of-dpa/controller.cpp]: should the 300 be configurable?

File: src/of-dpa/controller.cpp [hunk 3]
```diff
@@ -127,28 +122,8 @@ void controller::handle_packet_in(rofl::crofdpt &dpt, const rofl::cauxid &auxid,
...
-  case OFDPA_FLOW_TABLE_ID_ACL_POLICY:
-    send_packet_in_to_cpu(dpt, msg);
-    break;
-  default:
-    LOG(WARNING) << __FUNCTION__ << ": unexpected packet-in from table "
-                 << msg.get_table_id() << ". Dropping packet.";
-    break;
-  }
+  // all packets go up
+  send_packet_in_to_cpu(dpt, msg);
```
Comment (DFritzsche) [inline src/of-dpa/controller.cpp]: sometimes its cpu, sometimes kernel, are those the same?

File: src/of-dpa/controller.cpp [hunk 4]
```diff
@@ -158,27 +133,20 @@ void controller::handle_flow_removed(rofl::crofdpt &dpt,
...
-  }
-#endif
-
   switch (msg.get_table_id()) {
   case OFDPA_FLOW_TABLE_ID_BRIDGING:
     handle_bridging_table_rm(dpt, msg);
     break;
   default:
+    LOG(WARNING) << __FUNCTION__ << ": unhandled flow removal in table_id="
+                 << msg.get_table_id();
```
Comment (DFritzsche) [inline src/of-dpa/controller.cpp]: table_id should be casted to int to get useful logging

File: src/netlink/tap_manager.cpp [hunk 5]
```diff
@@ -163,35 +27,41 @@ int tap_manager::create_tapdev(uint32_t port_id, const std::string &port_name,
...
       int fd = dev->get_fd();
 
       LOG(INFO) << __FUNCTION__ << ": port_id=" << port_id
                 << " portname=" << port_name << " fd=" << fd << " ptr=" << dev;
 
-      // start reading from port
-      io.register_tap(fd, port_id, cb);
+      {
+        std::lock_guard<std::mutex> lock{tn_mutex};
+        tap_names2fds.emplace(std::make_pair(port_name, fd));
```
Comment (DFritzsche) [inline src/netlink/tap_manager.cpp]: tap_names2fds is neither cleared nor erased

File: src/netlink/cnetlink.cpp [hunk 6]
```diff
@@ -282,6 +300,114 @@ void cnetlink::nl_cb_v2(struct nl_cache *cache, struct nl_object *old_obj,
...
+  std::deque<nl_pkt_in> _packet_in;
+
+  {
+    std::lock_guard<std::mutex> scoped_lock(pi_mutex);
+    _packet_in.swap(packet_in);
+  }
+
+  for (int cnt = 0; cnt < nl_proc_max && _packet_in.size() && running; cnt++) {
+    auto p = _packet_in.front();
+    int ifindex = tap_man->get_ifindex(p.port_id);
```
Comment (DFritzsche) [inline src/netlink/cnetlink.cpp]: so, that's an `ifi` then ;)

File: src/netlink/cnetlink.cpp [hunk 7]
```diff
@@ -25,10 +25,20 @@
...
+cnetlink::cnetlink()
     : swi(nullptr), thread(this), caches(NL_MAX_CACHE, nullptr),
-      tap_man(tap_man), bridge(nullptr), nl_proc_max(10), running(false),
-      rfd_scheduled(false), l3(new nl_l3(tap_man, this)) {
+      bridge(nullptr), nl_proc_max(10), running(false), rfd_scheduled(false),
+      l3(new nl_l3(this)) {
+
+  sock_tx = nl_socket_alloc();
+  if (sock_tx == nullptr) {
+    LOG(FATAL) << "cnetlink: failed to create netlink socket" << __FUNCTION__;
```
[the remark this answers is not in the document]
Comment (toanju) [reply src/netlink/cnetlink.cpp]: historic reasons (i.e. did not touch that part)

File: src/netlink/cnetlink.cpp [hunk 8]
```diff
@@ -282,6 +300,114 @@ void cnetlink::nl_cb_v2(struct nl_cache *cache, struct nl_object *old_obj,
...
+    tap_man->enqueue(p.fd, p.pkt);
+
+    _packet_in.pop_front();
+  }
+
+  int size = _packet_in.size();
+  if (size) {
+    std::lock_guard<std::mutex> scoped_lock(pi_mutex);
+    std::copy(make_move_iterator(_packet_in.begin()),
+              make_move_iterator(packet_in.end()),
```
[the remark this answers is not in the document]
Comment (toanju) [reply src/netlink/cnetlink.cpp]: good catch, thanks

File: src/netlink/cnetlink.cpp [hunk 9]
```diff
@@ -282,6 +300,114 @@ void cnetlink::nl_cb_v2(struct nl_cache *cache, struct nl_object *old_obj,
...
+              std::back_inserter(packet_in));
```
[the remark this answers is not in the document]
Comment (toanju) [reply src/netlink/cnetlink.cpp]: true, completely forgot about this, thanks

File: src/netlink/cnetlink.cpp (hunk 1 above)
Comment (toanju) [reply src/netlink/cnetlink.cpp]: ah, one line below... fixed

File: src/of-dpa/controller.cpp (hunk 2 above)
Comment (toanju) [reply src/of-dpa/controller.cpp]: could, but it's pretty much fixed in the kernel as well hence I would do this as a follow up (later)

File: src/of-dpa/controller.cpp (hunk 3 above)
Comment (toanju) [reply src/of-dpa/controller.cpp]: It was just kernel in the 'comment' (fixed above). 'To CPU' points into the direction the packets go which in our case eventually will be the Linux kernel. Thus both would be correct though.

File: src/of-dpa/controller.cpp (hunk 4 above)
Comment (toanju) [reply src/of-dpa/controller.cpp]: unsigned ;)

File: src/netlink/tap_manager.cpp (hunk 5 above)
Comment (toanju) [reply src/netlink/tap_manager.cpp]: good catch, thanks

File: src/netlink/cnetlink.cpp (hunk 6 above)
Comment (toanju) [reply src/netlink/cnetlink.cpp]: :rofl:

File: src/netlink/tap_manager.cpp (hunk 5 above)
Comment (DFritzsche) [reply src/netlink/tap_manager.cpp]: we should add a `tap_names2fds.clear()` to here: https://github.com/toanju/basebox/blob/b985724583959f13a72ac9cad8e3e110d99635aa/src/netlink/tap_manager.cpp#L118

File: src/of-dpa/controller.cpp (hunk 3 above)
Comment (DFritzsche) [reply src/of-dpa/controller.cpp]: well, I like having one name for one thing, that's all

File: src/netlink/tap_manager.cpp (hunk 5 above)
Comment (toanju) [reply src/netlink/tap_manager.cpp]: solved slightly differntly

**Pull request merged**